include/wil/registry.h

```diff
@@ -17,16 +17,3139 @@
+                nullptr, nullptr, nullptr, nullptr, nullptr, nullptr));
+            return numSubKeys;
+        }
+
+        /**
+         * \brief Queries for number of values
+         * \param key The HKEY to query for number of values
+         * \return The queried number of value if succeeded
+         * \exception std::exception will be thrown on failure (including wil::ResultException)
+         */
+        inline size_t get_child_value_count(HKEY key)
+        {
+            DWORD numSubValues{};
+            THROW_IF_WIN32_ERROR(::RegQueryInfoKeyW(
+                key, nullptr, nullptr, nullptr, nullptr, nullptr,
+                nullptr, &numSubValues, nullptr, nullptr, nullptr, nullptr));
+            return numSubValues;
+        }
+#endif // #if defined(WIL_ENABLE_EXCEPTIONS)
+        /**
+         * \brief Queries for number of sub-keys
+         * \param key The HKEY to query for number of sub-keys
+         * \param[out] numSubKeys A pointer to a DWORD to receive the returned count
+         * \return HRESULT error code indicating success or failure
+         *         does not throw C++ exceptions
+         */
+        inline HRESULT get_child_key_count_nothrow(HKEY key, _Out_ DWORD* numSubKeys) WI_NOEXCEPT
+        {
+            RETURN_IF_WIN32_ERROR(::RegQueryInfoKeyW(
+                key, nullptr, nullptr, nullptr, numSubKeys, nullptr,
+                nullptr, nullptr, nullptr, nullptr, nullptr, nullptr));
+            return S_OK;
+        }
+
+        /**
+         * \brief Queries for number of values
+         * \param key The HKEY to query for number of sub-keys
+         * \param[out] numSubValues A pointer to a DWORD to receive the returned count
+         * \return HRESULT error code indicating success or failure
+         *         does not throw C++ exceptions
+         */
+        inline HRESULT get_child_value_count_nothrow(HKEY key, _Out_ DWORD* numSubValues) WI_NOEXCEPT
+        {
+            RETURN_IF_WIN32_ERROR(::RegQueryInfoKeyW(
+                key, nullptr, nullptr, nullptr, nullptr, nullptr,
+                nullptr, numSubValues, nullptr, nullptr, nullptr, nullptr));
+            return S_OK;
+        }
+
+#if defined(WIL_ENABLE_EXCEPTIONS)
+        /**
+         * \brief Writes a value under a specified key, the registry type based off the templated type passed as data
+         * \tparam T The type capturing the data being set
+         * \param key An opened registry key, or fixed registry key as the base key, from which to append the path
+         * \param subkey A string to append to the HKEY to attempt to write
+         *        can be nullptr if not needed
+         * \param value_name A string specifying the name of the registry value to write
+         *        can be nullptr to write to the unnamed default registry value
+         * \param data The value containing the data to be set in the specified key
+         *        note - is passed by const-ref

```

Comment:
This is aligned with my earlier comments about redundancy, but this is lanugage nuance visible in the function signature. We should not document it here.